x-pack/plugins/fleet/server/services/security/message_signing_service.ts

```diff
@@ -0,0 +1,142 @@
+
+interface MessageSigningKeys {
+  private_key: string;
+  public_key: string;
+  passphrase: string;
+}
+
+export interface MessageSigningServiceInterface {
+  getPublicKey(): Promise<string>;
+  sign(message: unknown): Promise<string>;
+  rotateKeys(passphrase: string): Promise<string>;
+}
+
+export class MessageSigningService implements MessageSigningServiceInterface {
+  constructor(
+    private soClient: SavedObjectsClientContract,
+    private esoClient: EncryptedSavedObjectsClient
+  ) {}
+
+  public async getPublicKey(): Promise<string> {
+    const { publicKey } = await this.#getCurrentKeyPair();
+    return publicKey;
+  }
+
+  public async sign(message: unknown): Promise<string> {
+    const signer = createSign('SHA256');
+    signer.update(JSON.stringify(message));
+    signer.end();
+
+    const { privateKey, passphrase } = await this.#getCurrentKeyPair();
+
+    return signer.sign({ key: privateKey, passphrase }, 'hex');
+  }
+
+  public async rotateKeys(passphrase?: string): Promise<string> {
+    return (await this.#generateKeyPair(passphrase)).publicKey;
+  }
+
+  async #getCurrentKeyPair(): Promise<{
+    privateKey: string;
+    publicKey: string;
+    passphrase: string;
+  }> {
+    const finder =
+      await this.esoClient.createPointInTimeFinderDecryptedAsInternalUser<MessageSigningKeys>({
+        type: MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE,
+        perPage: 1,
+        sortField: 'created_at',
+        sortOrder: 'desc',
+      });
+    let keyPair = {
+      privateKey: '',
+      publicKey: '',
+      passphrase: '',
+    };
+    for await (const result of finder.find()) {
+      const attributes = result.saved_objects[0]?.attributes;
+      if (!attributes?.private_key) {
+        break;
+      }
+      keyPair = {
+        privateKey: attributes.private_key,
+        publicKey: attributes.public_key,
+        passphrase: attributes.passphrase,
+      };
+      break;
+    }
+
+    // key pair not created yet, create now
+    if (!keyPair.privateKey) {
+      const generatedKeyPair = await this.#generateKeyPair();
+      keyPair = {
+        privateKey: generatedKeyPair.privateKey,
+        publicKey: generatedKeyPair.publicKey,
+        passphrase: generatedKeyPair.passphrase,
+      };
+    }
+
+    if (!keyPair.privateKey) {
+      throw new Error('failed to get private key');
+    }
+
+    if (!keyPair.publicKey) {
+      throw new Error('failed to get public key');
+    }
+
+    if (!keyPair.passphrase) {
+      throw new Error('failed to get passphrase');
+    }
+
+    return keyPair;
+  }
+
+  async #generateKeyPair(
+    providedPassphrase?: string
+  ): Promise<{ privateKey: string; publicKey: string; passphrase: string }> {
+    const passphrase = providedPassphrase || this.#generatePassphrase();
+    const keyPair = generateKeyPairSync('rsa', {
+      modulusLength: 4096,
+      privateKeyEncoding: {
+        type: 'pkcs8',
+        format: 'pem',
+        cipher: 'aes-256-cbc',
+        passphrase,
```

Comment:
encrypting the private key with a passphrase was in the acceptance criteria but it does feel a bit redundant here since the private key is already encrypted as a part of encrypted saved objects. if this is compromised, then the passphrase is also compromised. we can alternatively store the passphrase in a different encrypted saved object type but this only feels marginally more secure?